### DISCUSSION TITLE
Digital engineering

### DISCUSSION THREAD

**Discussion by iamdrscott** [niemopen/ntac-admin#41] [General]
**NIEM 6 digital engineering**

I think it will always be possible to generate by hand every artifact needed for a NIEM 6 message specification.  But that will not be the usual case.  Message designers will edit some artifacts, and then others will be generated by tools.  The generation of those artifacts is a good place for build automation.  I want to work through the steps of that build automation and ensure that CMFTool has a command-line interface for each step.  I also want to provide examples of how a NIEM message specification can fit into the build automation of software that produces or consumes those messages.

CMFTool will support both CMF and XSD model representations.  When you see a command line capability like "merge several CMF files into one CMF file", assume there will be a parallel capability like "merge several NIEM-conforming XSD schema document piles into one XSD schema document pile".

One objective is to support developing a message specification on a disconnected computer.  I want to burn all my tools plus the files for a NIEM model version onto a CDROM, take it into a classified lab, and design a message specification without having any internet access.

**A complex message specification**

A complex message specification defines two or more *message types*, each of which defines two or more *message formats*.  For example, the message types might be "request" and "response".  The "request" message type might define a canonical XML and JSON message format.  I imagine a directory structure convention like this:

```
MessageSpec
├── README.md                      *created by message designer*
├── manifest.xml                   *created by message designer*
├── type.request
│   ├── format.json
│   │   ├── context.json
│   │   └── validate.json
│   ├── format.xml
│   │   ├── transform.xslt
│   │   └── validate.xsd
│   │       ├── adapters
│   │       ├── domains
│   │       ├── external
│   │       ├── message.xsd
│   │       ├── niem-core.xsd
│   │       ├── utility
│   │       └── xml-catalog.xml
│   ├── messageModel.cmf           *created by message designer*
│   └── propertyMap.xml            *created by message designer*
└── type.response
    └── . . .
```

(We won't need all those directories in a simpler message specification.  If the specification defines only one message type, you don't need "type.foo" directories.  If a message type has only one format, you don't need '"format.foo" directories.)

**Artifacts created by the message designer**

I'm not going to talk about *README.md* and *manifest.xml* now.  I'll start with the message model.

At one point I thought the message designer would begin by creating a wantlist for a subset of a NIEM model version.  I no longer see a need for subset schema generation as a distinct step.  Instead I imagine the message designer using a model editor to create a message model for each message type in the specification.  The model editor loads the reference model from a NIEM model version, plus anything else the message designer wants to pick from.  The designer chooses components from the reference model creates extension components as required, and assembles them into the message model.  The editor produces a message model, plus a reference model for any extension components created (for the benefit of others).

My imaginary message editor has a left and right pane.  The left pane has a tree view of the message content.  The right pane allows search and display of reference model components.  The designer finds what he wants in the right pane and inserts it into the message model in the left pane, perhaps by drag-and-drop.  If the designer can't find what he wants in the reference model, then he creates a new component and inserts that into the message model.  The editor might look something like this:

<img width="601" alt="DigitalEngineering-fig1" src="https://user-images.githubusercontent.com/9931094/224995433-b781b63b-2cf4-4a70-be0c-701e7daafa0e.png">

(I built that mockup with JavaFX in an hour.  I think a working prototype could be built in a hundred hours or so.)

If we don't have a model editor, if it's necessary to generate a subset schema and then create an extension schema, then I'll need a CMFTool capability to generate a subset from a NIEM version plus a wantlist.

Once the message model is finished, if the message formats are to use canonical property names, then we're done.  If the designer wants a message format that uses simple property names, e.g. `lname` for `PersonSurName`, then he has to provide a property mapping file.  I'm not trying to figure out how to represent those mappings yet. (I want to see how COTS tools like Altova MapForce represent their mappings.)

**Artifacts to be created by build automation**

Still not sure whether this should be controlled by a standard automation tool (make, ant, maven, gradle), or a bespoke program reading the *manifest.xml* file.  Either way, any time the designer changes a message model or property mapping, the following artifacts need to be updated:

* *context.json* -- JSON-LD @context object
* *validate.json* -- JSON Schema
* *validate.xsd (usual)* -- XSD pile for validation & binding, not a model  (non-conforming)
* *validate.xsd (simple)* -- single XSD document for validation & binding (non-conforming)
* *transform.xslt* -- conversions to/from canonical NIEM XML

**CMFTool capabilities**

In order to generate the artifacts through build automation, CMFTool needs a command-line interface for these things:

* generate a reference schema (XSD) from CMF
* generate a message schema (XSD) from CMF
* generate CMF from a reference or message schema (XSD)
* generate JSON-LD @context from CMF
* generate JSON Schema from CMF
* generate a validation XSD pile and XSLT transforms from CMF and a property mapping
* generate a validation XSD document and XSLT transforms from CMF and a property mapping

**Comment by TomCarlson-NTAC** [comment]
Expect you rarely want the entire object. So there needs to be a way in the Message Editor to expand an object and only grab what you want inside it. And to do the same with the things inside it. And inside those. And now you have the SSGT.

Adding the ability to create objects inside the SSGT, or in its coming replacement, seems like a shorter path.

**Reply by iamdrscott** [reply]
Sure, that works for me!  I will be happy so long as I can:
1. Add components I need for my message model, and generate a reference model for them
2. Reuse components from any pile of reference models I choose, not just NIEM versions
3. Do all the work on my local machine with no network access

If the son-of-SSGT does that, great!  What I _don't_ want is my current workflow
1. Use SSGT to build a subset schema
2. Use Oxygen or whatever to build extension schema documents
3. Load my wantlist back into SSGT so I can add the things I forgot the first time
4. Use Oxygen or whatever to modify my extension schema
5. _repeat_

**Reply by TomCarlson-NTAC** [reply]
It think 1 & 2 are in the plan. We want to support the community adding content that can then be used by others.

Bullet 3? Not so much. We're building front-ends to the backend API to replace some existing tool functionality. That pretty much requires network access.

I would love some sort of new visual paradigm, but I haven't come up with one that's easier to use than the SSGT.

**Reply by iamdrscott** [reply]
Classified labs are pretty important for the two NIEMOpen premier sponsors.  They're going to want bullet 3.  So I'm puzzled.  Why not build something that can run on a local machine, and then put a web interface on top of that?   What's the advantage of the network-only approach?

**Reply by cdmgtri** [reply]
We will have to support being able to run it locally for the MOMS releases.

**Reply by TomCarlson-NTAC** [reply]
Well, you _could_ run it locally, maybe in a Docker container. That's what happens with MOMS now.

But we're talking the database that holds the model, the one where we build and maintain NIEM. And where the community could upload content for others to use.

You could build whatever front end you wanted for that, or use the API to build your own data source for a local tool w/o network access. That's just not what GTRI is doing via the NIEM API 2.0.
